_posts/2015-09-29-Asking-Users-for-Help.md

```diff
@@ -0,0 +1,155 @@
+
+- The thread names / ids on which the methods were executed
+ - This helps you identify race conditions.
+
+- The states of the user settings after a change
+
+- The data we got from the backend
+
+- The data we sent to the backend
+
+####Important things you should keep in mind when you create logs:
+- Log everything you need to identify your problem. If you forget something crucial you are in trouble! Once you send your version to the user, the only thing you can do, is create a new version. But how many times can you send a new version to a user, before he decides that he is getting very annoyed by you?
+
+- Don’t log too much. At first we logged so much information that it was hard to read what was going on. By reducing the information to the important parts it became a lot easier.
+
+- Choose nice and readable formatting. Invest a bit of time in this. The better the logs are formatted the better you will be able to process the information. If you want to be extra fancy think of some way you can automatically process your logs. But for us it was enough to read the logs the old fashion way.
+
+- Be careful with sensitive information! You don’t want to leak private information of the user in any way, so be careful what you log and how you send it around.
+
+##How to log
+
+We decided to use our bus system which is available throughout the app. Wherever we want to log something we post a log event with the relevant data. This log event looks like this:
+
+
+```java
+public class DebugLogEvent extends Event
+{
+   private static Clock sClock;
+   public final String methodName;
+   public final long threadId;
+   public final String threadName;
+   public final String message;
+   public long elapsedTime;
+   public long uptime;
+
+   public DebugLogEvent(String methodName, Object objectToLog)
+   {
+       Clock clock = sClock;
+       if(clock == null)
+       {
+           clock = new Clock();
+       }
+
+       this.elapsedTime = clock.getElapsedTime();
+       this.uptime = clock.getUptime();
+       this.methodName = methodName;
+       Thread currentThread = Thread.currentThread();
+       this.threadId = currentThread.getId();
+       this.threadName = currentThread.getName();
+       this.message = String.valueOf(objectToLog);
+   }
+
+   public static void setClock(Clock clock)
+   {
+       DebugLogEvent.sClock = clock;
+   }
+}
+```
+In the constuctor of the DebugLogEvent we set the method name which we formatted like this: “{className}.{methodName}”. The object we passed provided us with the log message through the toString() method. This way we could either pass in a string or any other object which had a well defined toString() method. As you can see all the other fields like the different timeStamps and thread names are automatically filled inside the constructor. The Clock class is just a wrapper which gives us the system time etc.
```

Comment:
You can make the code parts stand out a bit more using simple backquotes like this `{className}.{methodName}`.